tests/PHPUnit/Integration/Session/SessionAuthTest.php

```diff
@@ -0,0 +1,165 @@
+        $this->assertEquals(AuthResult::FAILURE, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsSuccess_IfRequestIpAndUserAgentMatchSession_AndCookieUsernameMatchesSessionUsername()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::SUCCESS, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsSuperUserSuccess_IfRequestIpAndUserAgentMatchSession_AndCookieUsernameMatchesSessionUsername()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::SUCCESS_SUPERUSER_AUTH_CODE, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsFailure_IfNoSessionExists()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, Fixture::ADMIN_USER_LOGIN);
+
+        $this->destroySession();
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::FAILURE, $result->getCode());
+    }
+
+    public function test_authenticate_ReturnsFailure_IfAuthenticatedSession_AndPasswordChangedAfterSessionCreated()
+    {
+        $this->initializeSession(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+        $this->initializeRequest(self::TEST_IP, self::TEST_UA, self::TEST_OTHER_USER);
+
+        sleep(1);
+
+        UsersManagerAPI::getInstance()->updateUser(self::TEST_OTHER_USER, 'testpass2');
+
+        $result = $this->testInstance->authenticate();
+        $this->assertEquals(AuthResult::FAILURE, $result->getCode());
+
+        $this->assertEmpty($_SESSION);
+    }
+
+    private function initializeRequest($ip, $userAgent, $userName, $passwordModifiedTime = false)
+    {
+        if (empty($passwordModifiedTime)) {
+            $model = new Model();
+            $user = $model->getUser($userName);
+            $passwordModifiedTime = $user['ts_password_modified'];
+
+            if (empty($passwordModifiedTime)) { // sanity check
+                throw new \Exception('ts_password_modified is empty');
+            }
+        }
+
+        $sessionCookieFactory = StaticContainer::get(SessionAuthCookieFactory::class);
+        $sessionFingerprint = new SessionFingerprint();
+
+        $cookie = $sessionCookieFactory->getCookie(false);
+        $cookie->set('id', $sessionFingerprint->getHash($passwordModifiedTime));
+
+        $cookieName = Config::getInstance()->General['login_cookie_name'];
+        $_COOKIE[$cookieName] = $cookie->generateContentString();
+
+        $_SERVER['REMOTE_ADDR'] = $ip;
```

Comment:
Have we checked IP's before for that? On mobile phones IP might often change...

Reply:
Associating sessions to a single IP/user agent/whatever is new. Its purpose is to prevent session hijacking by making it harder for stolen session cookies to be used by attackers. If there's some other information that can be used to identify the device/client/user that can't be faked, could use that. Otherwise I can just remove this from the PR.

Reply:
Yeah I know the reason it is there for and user agent should be definitely used. I just know that some people had problems in the past using IPs for this when they change regularly. Not sure what else would be save to use (have limited internet access)

Reply:
User Agent is good to have (now I think I know why I get logged out of some services when upgrading browser version :smile: )

IP address probably shouldn't be used as it could lead to frequent logouts in case people have rotating IPs every day or so.

Reply:
Here’s an idea: we can include a nonce in the cookie and change it on each request. If it doesn’t match what’s in the session, then the session is destroyed. If an attacker intercepts an in progress session, they will only be able to use it until the legit user sends another request. Not sure how well this will work with concurrent requests (though I think it might) and it doesn’t protect sessions that are idle for a long time (because of a large remember me value). What do you think?